R/cosh_nlsfit.R

```diff
@@ -0,0 +1,210 @@
+## sum_i (a_i cosh(m_i*t))
+## a_i are the amplitudes, m_i the masses and t is a vector of times
+sum.cosh <- function(masses, amplitudes, t){
+  colSums(amplitudes*cosh(masses%o%t))
+}
+
+## extract the effective mass from the sum of coshs
+cosh.to.effmass <- function(masses, amplitudes, t, Thalf, type){
+  if(type == "solve"){
+    sapply(t, FUN=function(x) {invcosh(sum.cosh(masses, amplitudes, x-Thalf)/sum.cosh(masses, amplitudes, x+1-Thalf), timeextent=2*Thalf, t=x+1)})
+  }else if(type == "acosh"){
+    acosh((sum.cosh(masses, amplitudes, t-Thalf-1)+sum.cosh(masses, amplitudes, t-Thalf+1))/(2*sum.cosh(masses, amplitudes, t-Thalf)))
+  }
+  ## the other types might be included as well...
+}
+
+## this fit-function works completely analogous to fit.effecitvemass
+## but it fits the correlator directly to a sum of cosh-terms
+fit.cosh <- function(effMass, cf, t1, t2, useCov=FALSE, m.init=0.01, par, n.cosh=2, every, ...){
+  if(missing(effMass) && missing(cf)){
+    stop("effMass or cf has to be provided!\n")
+  }
+  if(!missing(effMass)){
+    stopifnot(inherits(effMass, 'effectivemass'))
+    Thalf = effMass$Time/2
+    use.effmass <- TRUE
+  }else{
+    stopifnot(inherits(cf, 'cf_boot'))
```

Comment:
shouldn't there be also `stopifnot(inherits(cf, 'cf_meta'))`?

Reply:
Indeed, as `cf$Time` is accessed one needs the `cf_meta` mixin. This has been added in the latest state of Johann's branch.